src/lib/pthread_hook.cc

```diff
@@ -0,0 +1,54 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0. This product includes software
+// developed at Datadog (https://www.datadoghq.com/). Copyright 2021-Present
+// Datadog, Inc.
+
+#include "allocation_tracker.hpp"
+#include "savecontext.hpp"
+
+#include <dlfcn.h>
+#include <pthread.h>
+#include <tuple>
+
+namespace {
+using Args = std::tuple<void *(*)(void *), void *>;
+
+void *mystart(void *arg) {
+  {
+    ddprof::AllocationTrackerDisablerForCurrentThread disabler;
+    cache_stack_end();
+  }
+  Args *args = reinterpret_cast<Args *>(arg);
+  auto [start_routine, start_arg] = *args;
+  delete args;
+  return start_routine(start_arg);
+}
+
+} // namespace
+
+/** Hook pthread_create to cache stack end address just after thread start.
+ *
+ * The rationale is to fix a deadlock that occurs when user code in created
+ * thread calls pthread_getattr:
+ * - pthread_getattr takes a lock in pthread object
+ * - pthread_getattr itself does an allocation
+ * - AllocationTracker tracks the allocation and calls savecontext
+ * - savecontext calls pthread_getattr to get stack end address
+ * - pthread_getattr is reentered and attempts to take the lock again leading to
+ * a deadlock.
+ *
+ * Workaround is to hook pthread_create and call `cache_stack_end` to
+ * cache stack end address while temporarily disabling allocation profiling for
+ * current thread before calling user code.
+ * */
+int pthread_create(pthread_t *thread, const pthread_attr_t *attr,
```

Comment:
I love this mechanism.
I also think it will be helpful to identify thread lifetimes.